staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -62,10 +62,24 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
 	return a.namespaceScoped
 }
 
-func (customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
-}
-
-func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+func (a customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	// This applies defaults from the CRD schema.
+	// Validation errors, if any, are ignored. We do another validation pass here
+	// because we want defaulting to occur before the actual validation.
+	customResource := obj.(*unstructured.Unstructured)
+	customResourceValidator := a.validator.validator
+	result := customResourceValidator.Validate(customResource.UnstructuredContent())
```

Comment:
in case of an error (we see that inside `result`, don't we?), we should skip the defaulting. The second validation will catch the errors.

Reply:
even if `ApplyDefaults` ignores error, I would put an if around explicitly.

Reply:
Depends, right? If a required value with a default is missing, I'd expect an error before defaulting and no error after

Reply:
@liggitt can you give an example? cannot follow.

Reply:
say a property "foo" is both required and has a defined default:
```yaml
...
required: ["foo"]
properties:
  foo:
    default: "bar"
...
```

If a user submits an object without "foo" present, that should fail validation (because it is required). If defaulting is applied, "foo" would be set to "bar", and the object would then pass validation, right?

Reply:
The JSON schema validator will not report missing required fields if there is a default. So the first validation in the defaulting steps passes, then the `ApplyDefaults` step sets the default for the required field. Then it goes through the second validation and it passed too because the required field is set now. Moreover, the default value is validated against the spec (which is required by OpenAPI v3, but not JSON Schema). So this all sounds how it should be.

Reply:
> The JSON schema validator will not report missing required fields if there is a default. 

That's really unexpected... that means if I just ran validation, but did not apply defaults, I could persist data that did not conform to the spec?

Reply:
We reject defaults at the moment in CRD validation, until we add this defaulting logic.